worker/broker/broker.go

```diff
@@ -0,0 +1,163 @@
+	"github.com/juju/juju/api/common"
+	"github.com/juju/juju/api/provisioner"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/container"
+	"github.com/juju/juju/container/broker"
+	"github.com/juju/juju/core/instance"
+	"github.com/juju/juju/core/machinelock"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/worker"
+)
+
+// Config describes the dependencies of a Tracker.
+//
+// It's arguable that it should be called TrackerConfig, because of the heavy
+// use of model config in this package.
+type Config struct {
+	APICaller     base.APICaller
+	AgentConfig   agent.Config
+	MachineLock   machinelock.Lock
+	NewBrokerFunc func(broker.Config) (environs.InstanceBroker, error)
+}
+
+// Validate returns an error if the config cannot be used to start a Tracker.
+func (config Config) Validate() error {
+	if config.APICaller == nil {
+		return errors.NotValidf("nil APICaller")
+	}
+	if config.AgentConfig == nil {
+		return errors.NotValidf("nil AgentConfig")
+	}
+	if config.MachineLock == nil {
+		return errors.NotValidf("nil MachineLock")
+	}
+	if config.NewBrokerFunc == nil {
+		return errors.NotValidf("nil NewBrokerFunc")
+	}
+	return nil
+}
+
+// Tracker loads a broker, makes it available to clients, and updates
+// the broker in response to config changes until it is killed.
+type Tracker struct {
+	config   Config
+	catacomb catacomb.Catacomb
+	broker   environs.InstanceBroker
+}
+
+// NewTracker returns a new Tracker, or an error if anything goes wrong.
+// If a tracker is returned, its Broker() method is immediately usable.
+//
+// The caller is responsible for Kill()ing the returned Tracker and Wait()ing
+// for any errors it might return.
+func NewTracker(config Config) (*Tracker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	machineTag := config.AgentConfig.Tag().(names.MachineTag)
+	provisioner := provisioner.NewState(config.APICaller)
+	result, err := provisioner.Machines(machineTag)
+	if err != nil {
+		return nil, errors.Annotatef(err, "cannot load machine %s from state", machineTag)
+	}
+	if len(result) != 1 {
+		return nil, errors.Annotatef(err, "expected 1 result, got %d", len(result))
+	}
+	if errors.IsNotFound(result[0].Err) || (result[0].Err == nil && result[0].Machine.Life() == params.Dead) {
+		return nil, worker.ErrTerminateAgent
+	}
+	machine := result[0].Machine
+	// Expose SupportedContainers <- ([]instance.ContainerType, bool)
+	// if anything but LXD worker.ErrTerminateAgent
+	instanceContainers, determined, err := machine.SupportedContainers()
+	if err != nil {
+		return nil, errors.Annotate(err, "retrieving supported container types")
+	}
+	if len(instanceContainers) == 0 && !determined {
+		return nil, errors.Annotate(err, "no container types deterimined")
+	}
+	// we only work on LXD, so check for that.
+	for _, containerType := range instanceContainers {
+		if containerType != instance.LXD {
+			return nil, worker.ErrTerminateAgent
```

Comment:
I don't think you want TerminateAgent. I believe that means "stop jujud" entirely.
It may be that you just want to return nil error which says the "worker is done", no need to restart.